storage/src/lib.rs

```diff
@@ -0,0 +1,4984 @@
+    pub fn if_generation_not_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_generation_not_match".into(), vec![value.into()]);
+        self
+    }
+    /// Makes the operation conditional on whether the object's current
+    /// metageneration matches the given value.
+    pub fn if_metageneration_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_metageneration_match".into(), vec![value.into()]);
+        self
+    }
+    /// Makes the operation conditional on whether the object's current
+    /// metageneration does not match the given value.
+    pub fn if_metageneration_not_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_metageneration_not_match".into(), vec![value.into()]);
+        self
+    }
+    /// Resource name of the Cloud KMS key, of the form
+    /// projects/my-project/locations/global/keyRings/my-kr/cryptoKeys/my-key,
+    ///  that will be used to encrypt the object. Overrides the object
+    /// metadata's kms_key_name value, if any.
+    pub fn kms_key_name(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("kms_key_name".into(), vec![value.into()]);
+        self
+    }
+    /// Name of the object. Required when the object metadata is not
+    /// otherwise provided. Overrides the object metadata's name value, if
+    /// any. For information about how to URL encode object names to be path
+    /// safe, see Encoding URI Path Parts.
+    pub fn name(mut self, value: impl Into<String>) -> Self {
+        self.url_params.insert("name".into(), vec![value.into()]);
+        self
+    }
+    /// Apply a predefined set of access controls to this object.
+    pub fn predefined_acl(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("predefined_acl".into(), vec![value.into()]);
+        self
+    }
+    /// Set of properties to return. Defaults to noAcl, unless the object
+    /// resource specifies the acl property, when it defaults to full.
+    pub fn projection(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("projection".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request if the target bucket is
+    /// requester-pays bucket.
+    pub fn provisional_user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("provisional_user_project".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request. Required for Requester
+    /// Pays buckets.
+    pub fn user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("user_project".into(), vec![value.into()]);
+        self
+    }
+
+    pub async fn upload(
+        mut self,
+        media: impl Into<BytesReader>,
```

Comment:
`BytesReader` will limit what programmers can pass in here.  They can pass in a `Bytes`, `Response`, or `File`.  If they have some other bytes source that isn't one of these three then they have to read all the bytes and pass it in as Bytes, which requires reading the entire stream into memory.  The size of the content to upload may exceed the size of RAM.

Instead, could this function take a `Box<tokio::io::AsyncRead>` parameter?  Then, no matter what their byte source, they can implement `AsyncRead` for it if they haven't already.

I think that would also obviate some of the code in `Bytes.rs`.  `tokio::fs::File` already implements `AsyncRead`.  We could provide helper functions to convert `reqwest::Response` and `Bytes` to `AsyncRead`s.

Reply:
Let me stew on this some more over the weekend. This is for sure the hardest part of the API to get right. I agree that it feel like we could do something here to be even more permissive. I think a convince type is still nice for the common cases regardless though. Implementing AsyncRead on something is not super intuitive, to me at least, so having some helpers seems like a nice to have. 

If this was just an `Box<AsyncRead>` is makes it quite hard to use for types like `Vec<u8>`, `String`, or a `std::fs::File` which may be common for input read from another source.

Reply:
Maybe this type could just have another internal varient and impl a From<Box< AsyncRead >>?

Reply:
> Maybe this type could just have another internal varient and impl a From<Box< AsyncRead >>?

I think we'll end up with a lot less code if accept a `Box<AsyncRead>` instead of adding another internal variant.

Reply:
For `Vec<u8>`, `String`, and the like, we could provide user a wrapper
```
let readCursor = AsyncCursorReader::from(std::io::Cursor);
```

For `std::fs::File`, I'm not sure.  It would be doing synchronous reads from an asynchronous function.  Not sure how/if we should support that.

Reply:
I did some playing around this weekend and I think an <'a>AsyncRead + Unpin works well here. This seems to be the common construct for tokio::io::copy and the likes. As long as we have some good examples I think that works. I will update code. I still question on the download side though. I will start a new thread for that.

Reply:
Sounds good.